### PR TITLE
fix for product slug candidates and history

### PR DESCRIPTION
**Pull request by resool**
🐛 
after deleting product slug was updating with unix timestamp prefix but there was no change in friendly_id_slugs table which keeps history of slugs and has unique constrant on `["slug", "sluggable_type", "scope", "locale"]`
```
  create_table "friendly_id_slugs", force: :cascade do |t|
    t.string "slug", null: false
    t.bigint "sluggable_id", null: false
    t.string "sluggable_type", limit: 50
    t.string "scope"
    t.datetime "created_at", precision: nil
    t.datetime "deleted_at", precision: nil
    t.string "locale"
    t.index ["deleted_at"], name: "index_friendly_id_slugs_on_deleted_at"
    t.index ["locale"], name: "index_friendly_id_slugs_on_locale"
    t.index ["slug", "sluggable_type", "locale"], name: "index_friendly_id_slugs_on_slug_and_sluggable_type_and_locale"
    t.index ["slug", "sluggable_type", "scope", "locale"], name: "index_friendly_id_slugs_unique", unique: true
    t.index ["sluggable_id"], name: "index_friendly_id_slugs_on_sluggable_id"
    t.index ["sluggable_type"], name: "index_friendly_id_slugs_on_sluggable_type"
  end  
```

this leads to error when adding product with the same name as removed one

## change 1
since we are avoid using conditional unique constraints and there was test `it 'keeps the history when the product is destroyed'` i decided to update soft deleted `friendly_id_slugs.slug` with prefix `deleted-{id}_` since the is nothing more unique than id :)

## change 2
change `slug_candidates` for products. add 'deleted' prefix to slug for soft deleted records

## change 3
move slug logic from `models/spree/product.rb` to `models/spree/product/slugs.rb`. same with specs
 

## Summary by CodeRabbit

* **Refactor**
  * Centralized product slug management into a dedicated module for enhanced multilingual support and soft deletion.
  * Improved slug uniqueness, history tracking, and recovery after deletion.
* **Tests**
  * Added detailed tests covering slug behavior, localization, and edge cases.
  * Removed obsolete slug-related tests from the main product model specs.

### CODE REVIEW

File: core/app/models/spree/product/slugs.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    module Slugs
+      extend ActiveSupport::Concern
+
+      included do
+        extend FriendlyId
+        include Spree::TranslatableResourceSlug
+
+        translates :slug
+        friendly_id :slug_candidates, use: [:history, :slugged, :scoped, :mobility], scope: spree_base_uniqueness_scope, slug_limit: 255
+        acts_as_paranoid
```
Comment (damianlegawiec) [inline core/app/models/spree/product/slugs.rb]: @resool let's not move this outside of the main model, this is not slug specific logic

`acts_as_paranoid` should stay in product model main file

Comment (resool) [reply core/app/models/spree/product/slugs.rb]: it is required to use soft delete callbacks like `after_restore'`. and i guessing that updated default scope and .with_deleted also
`undefined method `after_restore' for class Spree::Product`

Comment (resool) [reply core/app/models/spree/product/slugs.rb]: however it would work with acts_as_paranoid defined on top of Product class (before includes) i guess 🤔

Comment (resool) [reply core/app/models/spree/product/slugs.rb]: yup. confirmed

File: core/app/models/spree/product/slugs.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        before_validation :downcase_slug
+        before_validation :normalize_slug, on: :update
+        after_destroy :punch_slugs
+        after_restore :regenerate_slug
+
+        validates :slug, presence: true, uniqueness: { allow_blank: true, case_sensitive: true, scope: spree_base_uniqueness_scope }
+
+        def slug_available?(slug, id)
+          !where(slug: slug).where.not(id: id).exists?
+        end
```
[the remark this answers is not in the document]
Comment (resool) [reply core/app/models/spree/product/slugs.rb]: this is already handled by acts_as_paranoid default scope

File: core/app/models/spree/product/slugs.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        translates :slug
+        friendly_id :slug_candidates, use: [:history, :slugged, :scoped, :mobility], scope: spree_base_uniqueness_scope, slug_limit: 255
+        acts_as_paranoid
+
+        Product::Translation.class_eval do
+          before_save :set_slug
+          acts_as_paranoid
+          # deleted translation values also need to be accessible for index views listing deleted resources
+          default_scope { unscope(where: :deleted_at) }
+
```
[the remark this answers is not in the document]
Comment (resool) [reply core/app/models/spree/product/slugs.rb:18]: copy pasted. leaving it as it was

File: core/spec/models/spree/product/slugs_spec.rb [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    product2.name = 'test'
+    product2.sku = '456'
+    product2.save!
+
+    expect(product2.slug).to eq 'test-456'
+  end
+
+  context 'history' do
+    before do
+      @product = create(:product, stores: [store])
```
Comment (Vegann) [inline core/spec/models/spree/product/slugs_spec.rb:117]: I know that this is an old spec, but since we are updating it, let's switch to `let` here
